crates/store/re_log_encoding/src/rrd/decoder/iterator.rs

```diff
@@ -0,0 +1,572 @@
+        let decoder = Self::new();
+        let mut it = DecoderIterator {
+            decoder,
+            reader,
+            wait_for_eos,
+            first_msg: None,
+        };
+
+        it.first_msg = it.next().transpose()?;
+
+        Ok(it)
+    }
+}
+
+// ---
+
+/// Iteratively decodes the contents of an arbitrary buffered reader.
+pub struct DecoderIterator<T, R: std::io::BufRead> {
+    decoder: Decoder<T>,
+    reader: R,
+
+    /// If true, the decoder will always wait for an end-of-stream marker before calling it a day,
+    /// even if the underlying reader has already reached its EOF state (…for now).
+    ///
+    /// This only really makes sense when running in tail mode (see `RetryableFileReader`),
+    /// otherwise we'd rather terminate early when a potentially short-circuited (and therefore
+    /// lacking a proper end-of-stream marker) RRD stream indicates EOF.
+    wait_for_eos: bool,
+
+    /// See [`Decoder::decode_eager`] for more information.
+    first_msg: Option<T>,
+}
+
+impl<T, R: std::io::BufRead> DecoderIterator<T, R> {
+    pub fn num_bytes_processed(&self) -> u64 {
+        self.decoder.byte_chunks.num_read() as _
+    }
+}
+
+impl<T: DecoderEntrypoint, R: std::io::BufRead> std::iter::Iterator for DecoderIterator<T, R> {
+    type Item = Result<T, DecodeError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(first_msg) = self.first_msg.take() {
+            // The iterator was eagerly initialized so make sure to return the first message if there's any.
+            return Some(Ok(first_msg));
+        }
+
+        loop {
+            match self.decoder.try_read() {
+                Ok(Some(msg)) => return Some(Ok(msg)),
+                Ok(None) => {}
+                Err(err) => return Some(Err(err)),
+            }
+
+            match self.reader.fill_buf() {
+                // EOF
+                Ok([]) => {
+                    // There's nothing more to read…
+                    match self.decoder.try_read() {
+                        // …but we still have enough buffered that we can still manage to decode
+                        // more messages, so go on for now.
+                        Ok(Some(msg)) => return Some(Ok(msg)),
+
+                        // …and we don't want to explicitly wait around for more to come, so just leave.
+                        Ok(None) if !self.wait_for_eos => return None,
+
+                        // …and the underlying decoder already considers that it's done (i.e. it's
+                        // waiting for a whole new stream to begin): time to stop.
+                        Ok(None) if self.decoder.state == DecoderState::StreamHeader => {
+                            return None;
+                        }
+
+                        // …but the underlying decoder doesn't believe it's done yet (i.e. it's still
+                        // waiting for an EOS marker to show up): we continue.
+                        Ok(None) => {}
```

Comment:
not super familiar with rust io. Is `fill_buf()` here the thing that guaranties the loop doesn't go 100% cpu nooping when in that situation?

Reply:
Yeah `fill_buf` lowers to standard IO under the hood